Rework and add new Blame-Diff plots

boehmseb opened the pull request:
This PR mainly reworks the b_correlation_matrix plot and adds a new b_distribution_comparison plot.
It also adds a plot that applies different clustering algorithms on blame-diff data.

boehmseb committed:
Add cluster algo comparison plot

boehmseb committed:
Ehnahce diff correlation matrix with logit and clustering

boehmseb committed:
Ehnahce diff correlation matrix with logit and clustering

boehmseb committed:
Unify plot generation between driver and artefacts

boehmseb committed:
Fix incorrect sort key in lorenz curve plot

boehmseb committed:
Add plot to compare distribution of variables

boehmseb committed:
Fix cluster analysis class name and description

boehmseb commented on varats/plots/blame_diff_correlation_matrix.py:
```diff
@@ -60,37 +146,136 @@ def plot(self, view_mode: bool) -> None:
+        df.drop(df[df.churn == 0].index, inplace=True)
+
+        _multivariate_grid(
+            x='churn',
+            y='num_interactions',
```
Any good ideas for making this configurable?
Should we use the extra-args mechanism of the plot-driver parser? Or generate multiple plots for all combinations?

vulder replied:
is this still relevant?

boehmseb replied:
Yes.

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 20
           

Complexity increasing per file
==============================
- varats/plots/cluster_analysis.py  8
- varats/plots/blame_diff_correlation_matrix.py  3
         

Complexity decreasing per file
==============================
+ varats/data/databases/blame_diff_metrics_database.py  -2
         

Clones added
============
- varats/plots/cluster_analysis.py  1
- varats/plots/blame_diff_correlation_matrix.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

boehmseb committed:
Merge branch 'vara-dev' into f-ClusterAnalysis

boehmseb committed:
Fix pylint warnings

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/plots/cluster_analysis.py  8
- varats/plots/blame_diff_correlation_matrix.py  3
         

Clones added
============
- varats/plots/cluster_analysis.py  1
- varats/plots/blame_diff_correlation_matrix.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

boehmseb committed:
add missing sklearn dependency

boehmseb committed:
Remove unused code

boehmseb committed:
Fix two more pylint warnings

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/cluster_analysis.py  8
- varats/plots/blame_diff_correlation_matrix.py  3
         

Clones added
============
- varats/plots/cluster_analysis.py  1
- varats/plots/blame_diff_correlation_matrix.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

vulder commented on varats/plots/cluster_analysis.py:
```diff
@@ -0,0 +1,217 @@
+class BlameDiffClusterAnalysis(Plot):
+    """Performs different cluster algorithms on blame-diff data."""
+
+    NAME = "cluster_analysis"
```
could be a bit more specific to prevent name clashes

vulder reviewed (approved):
looks good overall ;)

boehmseb committed:
Merge branch 'vara-dev' into f-ClusterAnalysis

boehmseb committed:
Merge branch 'f-UnifyPlotGeneration' into f-ClusterAnalysis

boehmseb committed:
Better handling of empty plots.

- always raise PlotDataEmpty when plot data is empty
- handle PlotDataEmpty consistently in Plot.save() and Plot.show()
- do not save/show empty plots and generate a warning
- close plots after saving/showing to prevent memory leaks when generating multiple plots

boehmseb committed:
Title for correlation matrix plot

boehmseb committed:
Fix bug in build_plot()

boehmseb committed:
Make distribution comparison plot axis variables configurable

boehmseb committed:
Rename cluster analysis plot and module

boehmseb committed:
Merge branch 'vara-dev' into f-ClusterAnalysis

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- varats/plots/blame_diff_correlation_matrix.py  3
- varats/plots/plot.py  1
- varats/plots/plots.py  7
- varats/plots/blame_diff_cluster_analysis.py  8
- varats/plots/commit_interactions.py  1
         

Complexity decreasing per file
==============================
+ varats/tools/driver_plot.py  -7
         

Clones added
============
- varats/plots/blame_diff_correlation_matrix.py  2
- varats/plots/paper_config_overview.py  1
- varats/plots/blame_diff_cluster_analysis.py  1
         

Clones removed
==============
+ varats/plots/blame_lorenz_curve.py  -2
+ varats/plots/repository_churn.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/blame_diff_correlation_matrix.py  3
- varats/plots/plot.py  1
- varats/plots/commit_interactions.py  1
- varats/plots/plots.py  1
- varats/plots/blame_diff_cluster_analysis.py  8
         

Clones added
============
- varats/plots/blame_diff_correlation_matrix.py  2
- varats/plots/paper_config_overview.py  1
- varats/plots/blame_diff_cluster_analysis.py  1
         

Clones removed
==============
+ varats/plots/blame_lorenz_curve.py  -2
+ varats/plots/repository_churn.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

boehmseb committed:
Fix diff_correlation_overview_table and its test

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/blame_diff_cluster_analysis.py  8
- varats/plots/plot.py  1
- varats/plots/blame_diff_correlation_matrix.py  3
- varats/plots/plots.py  1
- varats/plots/commit_interactions.py  1
         

Clones added
============
- varats/plots/blame_diff_cluster_analysis.py  1
- varats/plots/blame_diff_correlation_matrix.py  2
- varats/plots/paper_config_overview.py  1
         

Clones removed
==============
+ varats/plots/repository_churn.py  -1
+ varats/plots/blame_lorenz_curve.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)

boehmseb committed:
Add missing type annotations

vulder commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/blame_diff_cluster_analysis.py  8
- varats/plots/blame_diff_correlation_matrix.py  3
- varats/plots/plot.py  1
- varats/plots/plots.py  1
- varats/plots/commit_interactions.py  1
         

Clones added
============
- varats/plots/blame_diff_cluster_analysis.py  1
- varats/plots/paper_config_overview.py  1
- varats/plots/blame_diff_correlation_matrix.py  2
         

Clones removed
==============
+ varats/plots/blame_lorenz_curve.py  -2
+ varats/plots/repository_churn.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5844099&bid=19134603)
